src/mo_frontend/typing.ml

```diff
@@ -694,28 +708,30 @@ and check_con_env env at ce =
     error env at "M0156" "block contains expansive type definitions%s" msg
   end;
 
-and infer_inst env tbs typs at =
+and infer_inst env sort tbs typs at =
   let ts = List.map (check_typ env) typs in
   let ats = List.map (fun typ -> typ.at) typs in
   match tbs,typs with
   | {T.bound; sort = T.Scope; _}::tbs', typs' ->
     assert (List.for_all (fun tb -> tb.T.sort = T.Type) tbs');
     (match env.async with
+     | (C.AwaitCap c | C.AsyncCap c) when sort = T.Shared T.Query || sort = T.Shared T.Write || sort = T.Local ->
+        (T.Con(c,[])::ts, at::ats)
+     | (C.CompositeAwaitCap c | C.CompositeCap c) when sort = T.Shared T.Query || sort = T.Shared T.Composite ->
+        (T.Con(c,[])::ts, at::ats)
      | C.ErrorCap
      | C.QueryCap _
-     | C.NullCap ->
+     | C.NullCap
+     | _ ->
         error env at "M0047"
           "send capability required, but not available (need an enclosing async expression or function body)"
```

Comment:
Probably, the error message is not very specific if the call between the modes update/query/composite query is not supported.

Reply:
Yes, I was considering revising those (maybe splitting send capability into update query and composite query capability.
Maybe this is the right time to do it.

Reply:
This is nothing important, could be also done later. I guess we can do that as part of an error message improvement work (together as a team across the entire compiler).